### PR TITLE
devel/scheduler: update intro and scheduling algo

### PR DESCRIPTION
**Pull request by hongchaodeng**
This is to update scheduler.md to make it easier for new dev to understand:
- extend "exploring the code" section and put it before "scheduling algorithm"
- update the scheduling algorithm diagram. The previous one is really confusing.

### CODE REVIEW

**Comment by hongchaodeng** [comment]
@timothysc @davidopp 
As a side note, docs are moved here.
We need a way to forward docs/devel/scheduler.md in main repo to here.

**Comment by davidopp** [comment]
Yes, please ask the contribex SIG when they are going to do that, as they are responsible for the move.

**Comment by hongchaodeng** [comment]
Assigning @timothysc @davidopp as we discussed in 1.6 code health plan.
I will incrementally update the docs.

**Comment by fabiand** [comment]
:+1: as a newbie this documentation update looks nice to me. Especially the new ascii art should get a place on my wall. My 2ct.

**Comment by timothysc** [comment]
I'll assign @jayunit100 b/c he last updated the doc.

File: contributors/devel/scheduler.md [hunk 1]
```diff
@@ -1,39 +1,63 @@
 # The Kubernetes Scheduler
 
-The Kubernetes scheduler runs as a process alongside the other master
-components such as the API server. Its interface to the API server is to watch
-for Pods with an empty PodSpec.NodeName, and for each Pod, it posts a Binding
-indicating where the Pod should be scheduled.
+The Kubernetes scheduler runs as a process alongside the other master components such as the API server.
+Its interface to the API server is to watch for Pods with an empty PodSpec.NodeName,
+and for each Pod, it posts a Binding indicating where the Pod should be scheduled.
```
Comment (jayunit100) [inline contributors/devel/scheduler.md]: binding

File: contributors/devel/scheduler.md [hunk 2]
```diff
@@ -1,39 +1,63 @@
...
-## The scheduling process
+## Exploring the code
+
+We are dividng scheduler into three layers from high level:
+- [plugin/cmd/kube-scheduler/scheduler.go](http://releases.k8s.io/HEAD/plugin/cmd/kube-scheduler/scheduler.go):
+  This is the main() entry that does initialization before calling the scheduler framework.
+- [pkg/scheduler/scheduler.go](http://releases.k8s.io/HEAD/pkg/scheduler/scheduler.go):
+  This is the scheduler framework that handles stuff (e.g. binding) beyond the scheduling algorithm.
+- [pkg/scheduler/generic_scheduler.go](http://releases.k8s.io/HEAD/pkg/scheduler/generic_scheduler.go):
+  This is the scheduling algorithm that basically assigns nodes for pods.
```
Comment (jayunit100) [inline contributors/devel/scheduler.md]: The scheduling algorithm that assigns nodes for pods.

File: contributors/devel/scheduler.md [hunk 3]
```diff
@@ -1,39 +1,63 @@
...
+    +-------------------+-------------------------+
+
+    Pred. filters: node 3 doesn't have enough resource
+
+    +-------------------+-------------------------+
+                        |
+                        |
+                        v
+    +-------------------+-------------------------+
+    |                left nodes:                  |
```
Comment (jayunit100) [inline contributors/devel/scheduler.md]: change "left nodes" to "remaining nodes" ..

**Comment by jayunit100** [comment]
thanks @hongchaodeng couple minor comments, otherwise lg !

File: contributors/devel/scheduler.md [hunk 4]
```diff
@@ -1,39 +1,63 @@
 # The Kubernetes Scheduler
 
-The Kubernetes scheduler runs as a process alongside the other master
-components such as the API server. Its interface to the API server is to watch
-for Pods with an empty PodSpec.NodeName, and for each Pod, it posts a Binding
-indicating where the Pod should be scheduled.
+The Kubernetes scheduler runs as a process alongside the other master components such as the API server.
+Its interface to the API server is to watch for Pods with an empty PodSpec.NodeName,
```
Comment (jayunit100) [inline contributors/devel/scheduler.md]: it's (not your fault, it existed before :) but might as well fix now right :)

**Comment by hongchaodeng** [comment]
@jayunit100 
Didn't see any comments :)

**Comment by jayunit100** [comment]
just added

File: contributors/devel/scheduler.md (hunk 4 above)
Comment (hongchaodeng) [inline contributors/devel/scheduler.md]: Can you read again? It should be "Its interface to the ... **is** to".

**Comment by jayunit100** [comment]
@hongchaodeng (1) fix the one other `Its` comment and (2) rebase to one commit, then lets and thanks agin

File: contributors/devel/scheduler.md (hunk 4 above)
Comment (hongchaodeng) [inline contributors/devel/scheduler.md]: @jayunit100 
I see no problem in this sentence:

"Its interface to the API server is to watch..."

Can you clarify what it should be?

**Comment by hongchaodeng** [comment]
Thanks @jayunit100 for review.
(1) please help answer my question.
(2) done

**Comment by davidopp** [comment]
@wojtek-t can you please review this?

**Comment by hongchaodeng** [comment]
Thanks @jayunit100 @davidopp .
I will merge it by tomorrow if no further question.

**Pull request merged**